Sources/Sails/Application.swift

```diff
@@ -110,24 +115,29 @@ class FunctionHandler: ChannelInboundHandler {
 
     private func send(_ response: FutureResponse, via context: ChannelHandlerContext) {
         var buffer = context.channel.allocator.buffer(capacity: 0)
-        response.response(on: context.eventLoop).whenSuccess { response in
-            let contentLength = try! response.content.encode(to: &buffer)
+        response.response(on: context.eventLoop).whenSuccess { [weak self] response in
+            guard let self = self else {
+                return
+            }
 
+            let contentLength = try! response.content.encode(to: &buffer)
             var headers = response.head.headers
             headers.add(name: "Content-Length", value: String(contentLength))
 
-            let head = HTTPServerResponsePart.head(
-                HTTPResponseHead(
+            let head = HTTPResponseHead(
                     version: HTTPVersion(major: 1, minor: 1),
                     status: response.head.status,
                     headers: headers
-                )
             )
 
-            _ = context.write(self.wrapOutboundOut(head))
+            context.write(self.wrapOutboundOut(.head(head)), promise: nil)
+            context.write(self.wrapOutboundOut(.body(.byteBuffer(buffer))), promise: nil)
 
-            context.writeAndFlush(self.wrapOutboundOut(.body(.byteBuffer(buffer)))).whenComplete { (_) in
-                _ = context.close()
+            let keepAlive = self.requestHead?.isKeepAlive == true
+            context.writeAndFlush(self.wrapOutboundOut(.end(nil))).whenComplete { _ in
+                if !keepAlive {
+                    context.close(promise: nil)
+                }
```

Comment:
Preferably all of these changes would be in a separate commit (looks like you pulled this stuff in from my other branch)

Reply:
Yea, unfortunately the tests wouldn't pass without them so I wasn't sure what else to do.

Reply:
Should I rip these out and maybe put them up in their own PR?

Reply:
I merged my PR with these changes. You can pull this stuff out and then rebase on `master` and you should be good